connect/json/src/main/java/org/apache/kafka/connect/json/JsonConverter.java

```diff
@@ -185,89 +185,92 @@ public Object convert(Schema schema, JsonNode value) {
-        TO_JSON_LOGICAL_CONVERTERS.put(Date.LOGICAL_NAME, new LogicalTypeConverter() {
+        LOGICAL_CONVERTERS.put(Time.LOGICAL_NAME, new LogicalTypeConverter() {
             @Override
-            public Object convert(Schema schema, Object value) {
+            public JsonNode toJson(final Schema schema, final Object value, final JsonConverterConfig config) {
                 if (!(value instanceof java.util.Date))
-                    throw new DataException("Invalid type for Date, expected Date but was " + value.getClass());
-                return Date.fromLogical(schema, (java.util.Date) value);
+                    throw new DataException("Invalid type for Time, expected Date but was " + value.getClass());
+                return JsonNodeFactory.instance.numberNode(Time.fromLogical(schema, (java.util.Date) value));
             }
-        });
 
-        TO_JSON_LOGICAL_CONVERTERS.put(Time.LOGICAL_NAME, new LogicalTypeConverter() {
             @Override
-            public Object convert(Schema schema, Object value) {
-                if (!(value instanceof java.util.Date))
-                    throw new DataException("Invalid type for Time, expected Date but was " + value.getClass());
-                return Time.fromLogical(schema, (java.util.Date) value);
+            public Object toConnect(final Schema schema, final JsonNode value) {
+                if (!(value.isInt()))
+                    throw new DataException("Invalid type for Time, underlying representation should be integer but was " + value.getNodeType());
+                return Time.toLogical(schema, value.intValue());
             }
         });
 
-        TO_JSON_LOGICAL_CONVERTERS.put(Timestamp.LOGICAL_NAME, new LogicalTypeConverter() {
+        LOGICAL_CONVERTERS.put(Timestamp.LOGICAL_NAME, new LogicalTypeConverter() {
             @Override
-            public Object convert(Schema schema, Object value) {
+            public JsonNode toJson(final Schema schema, final Object value, final JsonConverterConfig config) {
                 if (!(value instanceof java.util.Date))
                     throw new DataException("Invalid type for Timestamp, expected Date but was " + value.getClass());
-                return Timestamp.fromLogical(schema, (java.util.Date) value);
+                return JsonNodeFactory.instance.numberNode(Timestamp.fromLogical(schema, (java.util.Date) value));
+            }
+
+            @Override
+            public Object toConnect(final Schema schema, final JsonNode value) {
+                if (!(value.isIntegralNumber()))
+                    throw new DataException("Invalid type for Timestamp, underlying representation should be integral but was " + value.getNodeType());
+                return Timestamp.toLogical(schema, value.longValue());
             }
         });
     }
 
-
-    private boolean enableSchemas = JsonConverterConfig.SCHEMAS_ENABLE_DEFAULT;
-    private int cacheSize = JsonConverterConfig.SCHEMAS_CACHE_SIZE_DEFAULT;
```

Comment:
I looked into the code for `AbstractConfig` and unless I'm missing something, it seems to store the deserialized config value, but I suppose we're still hitting the map lookup and the auto-unboxing for primitives. I'm skeptical that it makes a noticeable difference - especially in the context of converting to/from JSON, which involves delegating to Jackson [de]serializers that are themselves pretty inefficient - but I'll change it back to the old code (adopting your second suggestion) anyway just to be safe. 

As a side note about inlining, have we ever run into a performance issue where that made a noticeable difference? I'd love to learn about that!

Reply:
@agavra instead of adding the `CachedConfigs` class could we perform the parsing at instation time in the `JsonConverterConfig` class and store results there, exposing them through getters? For example:

```java
public class JsonConverterConfig extends AbstractConfig {

    private final DecimalFormat decimalFormat;

    // ...

    public JsonConverterConfig(Map<String, ?> props) {
        super(CONFIG, props);
        this.decimalFormat = DecimalFormat.valueOf(getString(DECIMAL_FORMAT_CONFIG).toUpperCase(Locale.ROOT));
    }

    // ...

    public DecimalFormat decimalFormat() {
        return decimalFormat;
    }
}
```

The `CachedConfigs` class adds a bit of cognitive overhead to this part of the code base; if we can avoid adding another utility class to have to track it'd be nice.

Reply:
@C0urante - I'm happy to do that; @rhauch laid out both options and had slight preference at caching them directly in the JsonConverter, but I can get behind the cognitive overhead argument. I'll update it to cache in the config unless he has any objections.

Reply:
My concern was that, unlike most of the places where we check configs, several of these checks are used in the (de)serialization methods that are called with every record. Prior to this change, those configs were cached inside the converter instance and not cached by the `JsonConverterConfig` class, and my concern was that we were slowing the overall performance of the (de)serialization with the multiple checks.

I think it's fine to cache them as final members in the `JsonConverterConfig` class, and reference them here, which is exactly what the current PR does.